public/js/client.js

```diff
@@ -17,53 +26,48 @@ function connectToClient(onConnect) {
   }).catch(err => console.log(err));
 }
 
-function connectToTab(tab, onNewSource) {
-  let deferred = promise.defer();
-  const options = { client: currentClient, form: tab, chrome: false };
-
-  TargetFactory.forRemoteTab(options).then(target => {
-    target.activeTab.attachThread({}, (res, threadClient) => {
-      threadClient.resume();
-      window.gThreadClient = threadClient;
-      currentThreadClient = threadClient;
-      deferred.resolve();
+function connectThread(tab, onNavigate) {
+  return new Promise((resolve, reject) => {
+    getTabTarget(tab).then(target => {
+      target.activeTab.attachThread({}, (res, threadClient) => {
+        threadClient.resume();
+        currentThreadClient = threadClient;
+        resolve();
+      });
     });
   });
-
-  return deferred.promise;
 }
 
-function getThreadClient() {
-  return currentThreadClient;
-}
+function debugTab(tab, actions) {
+  return Task.spawn(function* () {
+    yield connectThread(tab);
+    actions.selectTab({ tabActor: tab.actor });
+
+    const target = yield getTabTarget(tab);
+    target.on("will-navigate", actions.willNavigate);
+    target.on("navigate", actions.navigate);
 
-function debugTab({ tabActor, newSource, paused, resumed,
-                    selectTab, selectSource, loadSources }) {
-  function listenToClient() {
-    let deferred = promise.defer();
     let client = getThreadClient();
 
     client.addListener("paused", (_, packet) => {
-      paused(packet);
+      actions.paused(packet);
       if (packet.why.type != "interrupted") {
```

Comment:
does it make sense for `selectSource` to move to the `paused` action? I say that because `paused` already dispatches a loadFrames action.